Quadtree

gingeralesy opened the pull request:
Quadtrees to allow keeping track of objects with a location and a size in a 2D plane. Reorders itself automatically to allow quickly fetching all the objects in the requested region.

It is based on bvh2's interface, with a number of searching and callback functions added.

As an example a test case where I had 10000 objects into a 5000x5000 space and ran updates to them where they moved a random amount.
```
10000 insertions in 0.063s
10000 updates in 0.075s
10000 removals in 0.065s
Search of a 1000x1000 area took 0.000s
```
I think the reason for updates to take longer is that it easily causes reordering, but the combination of clearing the tree and then re-inserting everything wasn't particularly faster. Insertion speeds of course depend largely on how much they cluster. More they do the worse it is.

Shinmera commented:
Looks good to me, style is nice, easy to read. I'd be fine with merging it.

My only comment at this point is that I see a worrying amount of copying going on in various functions. I can't say whether that's going to be bad for GC hiccups, but it's something that should be looked at for future revisions. Quite a few of the functions could also safely be declared inline since they're tiny / have their types declared for better inference.

Shinmera commented:
Another feature that would be nice to have is an auto-resizer to reshape the tree if objects are added or moved out of bounds.

gingeralesy commented:
Alright, first there are some refactorings done. Region overlap/containment checks are named region-*, safety declarations are gone, instead of using append nconc is used, and I added inlining where they felt applicable.

Secondly there is now support for FOR iterator. Supports using it for every contained object and for a specific region with `(for:for ((object over quadtree :region (vec4 0 0 800 600))) ...)`. The callback functions use this now as well.

Thirdly there was quite a big bug that made the re-ordering of tree content not work after a removal. I wondered about this because removing everything was so fast. Running the tests again.
```
10000 insertions in 0.022
10000 updates in 0.106
Search took 0.000
10000 reinserts in 0.028
10000 removals in 0.161
````
* Updates included moving the object around slightly, so it didn't trigger a full re-insert every time.
* Re-insert did a full clear, updated every object's position slightly, and then re-inserted them.
* Removals take _much_ longer than this (seconds!) in trees that cover a particularly large area even if the object count is the same.
  * It's because a removal causes restructuring within the tree easily that then snowballs deeper into it.
  * In particular if a near root-level stored object is removed.
  * In those cases it's much better to just clear the entire tree and re-insert them since the child-nodes already exist.
  * Though it's mostly the case if there's a lot of clustering, like everything is in the one hundreth of the entire region.

Also as a note, the requested feature of the quadtree's covered area expanding if something outside of it is inserted was already there. The `QUADTREE-INSERT` calls `NODE-INSERT-EXTEND` which calls `NODE-EXTEND` when necessary and creates a new root node for the quadtree.

Shinmera commented:
Cool, thanks a lot for the changes!

If you want to optimise the consing more I'd change from using lists to an API that allows passing an adjustable vector as an argument, which it then pushes the matching elements into. That way people can recycle an existing vector and avoid consing altogether.

gingeralesy commented:
Made the suggested change of filling objects into vectors rather than creating a lot of conses.

Also there was a bug caused by a typo in the iterator and I ended up almost refactoring the entire FOR:STEP-FUNCTIONS before I found it. Tests run fine still and I coludn't find any more issues.
